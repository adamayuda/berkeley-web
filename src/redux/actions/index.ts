import { Dispatch } from "redux";
import axios from "axios";
import { mainStore } from "src/redux/index";

export const fetchPosts = () => async (
  dispatch: Dispatch,
  getState: typeof mainStore.getState,
) => {
  const { data: authInfo } = await axios.post<{
    user: {
      username: string;
      picture: string;
    };
    token: string;
  }>(
    "https://berkeley.cashcashbangbang.com/auth/login",
    {
      username: "berkeley",
    },
    {
      headers: {
        "Content-Type": "application/json",
      },
    },
  );
  const {
    user: { username, picture },
    token,
  } = authInfo;
  dispatch({
    type: "UPDATE_USER_INFO",
    payload: { username, picture, token },
  });

  const { data: posts } = await axios.get<{
    posts: [{ likes: number; _id: string; user: string; createdAt: Date }];
  }>("https://berkeley.cashcashbangbang.com/posts", {
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${getState().user.token}`,
    },
  });

  dispatch({
    type: "UPDATE_POSTS",
    payload: posts,
  });
};
