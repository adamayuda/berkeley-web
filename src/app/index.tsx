import "./style.sass";

import React, { FC, Suspense, lazy, useEffect } from "react";

import { Route, Switch } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";

import { Loading } from "src/components/loading";
import { StateInterface } from "src/redux";

const routes = [
  {
    import: import("src/pages/home"),
    path: "/",
    exact: true,
  },
];

export const App: FC = () => {
  const state = useSelector<StateInterface, StateInterface>((state) => state);
  const dispatch = useDispatch();

  const DEVICE_SCREEN_SIZE = 975;

  const onResize = () => {
    if (
      window.innerWidth <= DEVICE_SCREEN_SIZE &&
      state.device.type === "laptop"
    )
      dispatch({
        type: "UPDATE_DEVICE_TYPE",
        payload: { type: "mobile" },
      });
    else if (
      window.innerWidth > DEVICE_SCREEN_SIZE &&
      state.device.type === "mobile"
    ) {
      dispatch({
        type: "UPDATE_DEVICE_TYPE",
        payload: { type: "laptop" },
      });
    }
  };

  const fetchUserInfo = () => {
    fetch("https://jsonplaceholder.typicode.com/users/1", {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    })
      .then((response) => response.json())
      .then((result) => {
        const { username } = result;
        dispatch({
          type: "UPDATE_USER_INFO",
          payload: { username, picture: "ok" },
        });
      })
      .catch((error) => {
        console.error("Error:", error);
      });
  };

  useEffect(() => {
    onResize();
    window.addEventListener("resize", onResize);

    fetchUserInfo();

    return function cleanup() {
      window.removeEventListener("resize", onResize);
    };
  }, [dispatch]);

  return (
    <Suspense fallback={<Loading />}>
      <Switch>
        {routes.map((route, index: number) => (
          <Route
            {...route}
            key={`route-${index}`}
            component={lazy(() => route.import)}
          />
        ))}
      </Switch>
    </Suspense>
  );
};
