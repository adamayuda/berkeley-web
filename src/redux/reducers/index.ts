import { combineReducers } from "redux";
import { device } from "./device";
import { user } from "./user";

export const mainReducer = combineReducers({
  device,
  user,
});
