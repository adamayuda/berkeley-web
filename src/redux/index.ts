import { compose, createStore } from "redux";
import thunk, { ThunkDispatch } from "redux-thunk";
import { applyMiddleware } from "redux";

import { mainReducer } from "src/redux/reducers";

const composeEnhancers =
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const mainStore = createStore(
  mainReducer,
  composeEnhancers(applyMiddleware(thunk)),
);

export const createMainStore = () => {
  return createStore(mainReducer, composeEnhancers(applyMiddleware(thunk)));
};

export type StateInterface = ReturnType<typeof mainStore.getState>;

export type ActionType = "UPDATE_DEVICE_TYPE" | "UPDATE_USER_INFO";

export interface Action<T> {
  type: ActionType;
  payload: Partial<T>;
}

export type Dispatch<A> = ThunkDispatch<
  StateInterface,
  Record<string, unknown>,
  Action<A>
>;
