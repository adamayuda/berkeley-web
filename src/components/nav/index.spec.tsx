import { cleanup, render } from "@testing-library/react";
import { Nav } from ".";
import { Provider } from "react-redux";
import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { createMainStore } from "src/redux";

let mainStore = createMainStore();

describe("src/components/nav/index.tsx", () => {
  afterEach(() => cleanup());
  beforeEach(() => {
    mainStore = createMainStore();
  });

  test("Render Nav component", () => {
    const renderedNav = render(
      <Provider store={mainStore}>
        <Router>
          <Nav />
        </Router>
      </Provider>,
    );
    expect(renderedNav.container).toMatchSnapshot();
  });

  test("Render Theme component when darkMode is false", () => {
    mainStore.dispatch({
      type: "UPDATE_DEVICE_TYPE",
      payload: { type: "laptop" },
    });

    const renderedTheme = render(
      <Provider store={mainStore}>
        <Router>
          <Nav />
        </Router>
      </Provider>,
    );
    expect(renderedTheme.container).toMatchSnapshot();
  });

  test("Render Theme component when darkMode is false", () => {
    mainStore.dispatch({
      type: "UPDATE_DEVICE_TYPE",
      payload: { type: "mobile" },
    });

    const renderedTheme = render(
      <Provider store={mainStore}>
        <Router>
          <Nav />
        </Router>
      </Provider>,
    );
    expect(renderedTheme.container).toMatchSnapshot();
  });
});
