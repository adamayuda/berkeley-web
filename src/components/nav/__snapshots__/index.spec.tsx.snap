// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/nav/index.tsx Render Nav component 1`] = `
<div>
  <nav>
    <div
      class="container"
    >
      <div
        class="navGrid"
      >
        <img
          alt="LinkedTeam"
          src="test-file-stub"
        />
        <img
          alt="LinkedTeam"
          src="test-file-stub"
        />
        <img
          alt="LinkedTeam"
          src="test-file-stub"
        />
      </div>
    </div>
  </nav>
</div>
`;

exports[`src/components/nav/index.tsx Render Theme component when darkMode is false 1`] = `
<div>
  <nav>
    <div
      class="container"
    >
      <div
        class="navGrid"
      >
        <img
          alt="LinkedTeam"
          src="test-file-stub"
        />
        <img
          alt="LinkedTeam"
          src="test-file-stub"
        />
      </div>
    </div>
  </nav>
</div>
`;

exports[`src/components/nav/index.tsx Render Theme component when darkMode is false 2`] = `
<div>
  <nav>
    <div
      class="container"
    >
      <div
        class="navGrid"
      >
        <img
          alt="LinkedTeam"
          src="test-file-stub"
        />
        <img
          alt="LinkedTeam"
          src="test-file-stub"
        />
        <img
          alt="LinkedTeam"
          src="test-file-stub"
        />
      </div>
    </div>
  </nav>
</div>
`;
